main.py

```diff
@@ -144,14 +145,96 @@ async def request_termination_of_experiment(rid: int):
     remote.request_termination(rid)
 
 
+def add_tracking_line(stmt_list: List[ast.stmt]) -> List[ast.stmt]:
+    """Modifies the given statements list to leave rtio logs that track each line.
+
+    Adds logs containing the corresponding line number before each line, recursively.
+
+    Args:
+        stmt_list: The list of ast statements.
+
+    Returns:
+        The list of modified ast statements.
+    """
+    modified_stmt_list = []
+    for stmt in stmt_list:
+        tracking_line_code = f"rtio_log('line', {stmt.lineno})"
+        tracking_line_stmt = ast.parse(tracking_line_code).body
+        modified_stmt_list.extend(tracking_line_stmt)
+        for attr in ("body", "handlers", "orelse", "finalbody"):
+            if hasattr(stmt, attr):
+                setattr(stmt, attr, add_tracking_line(getattr(stmt, attr)))
+        modified_stmt_list.append(stmt)
+    return modified_stmt_list
+
+
+def modify_experiment_code(code: str, experiment_cls_name: str) -> str:
+    """Modifies the given code to leave rtio logs and make the vcd file.
+
+    It performs the following modifications:
+        1. Add the rtio logs to track each line.
+        2. Make the vcd file into the corresponding path.
+
+    Args:
+        code: The experiment code.
+        experiment_cls_name: The class name of the experiment.
+
+    Returns:
+        The modified experiment code.
+    """
+    code_mod = ast.parse(code)
+    # import modules
+    import_code = """
+import subprocess
+import posixpath
+    """
+    import_stmt = ast.parse(import_code).body
+    code_mod.body = import_stmt + code_mod.body
+    experiment_cls_stmt = next(
+        stmt for stmt in code_mod.body
+        if isinstance(stmt, ast.ClassDef) and stmt.name == experiment_cls_name
+    )
+    run_func_stmt = next(
+        stmt for stmt in experiment_cls_stmt.body
+        if isinstance(stmt, ast.FunctionDef) and stmt.name == "run"
+    )
```

Comment:
It will rarely happen if these conditions are not met, so the docsting is likely to be sufficient.

Reply:
Okay! An user will submit the experiment through IQUIP and the GUI checks if the selected file is a valid experiment, i.e. if it contains the experiment class.
Thus, the experiment class always exists.

The problem is that `run()` doesn't exist, but this will not happen unless the user intends it, so I will only mark it in the docstring.

Reply:
Good! As you said, to guarantee such condition, all the other parts should work well and the mechanism should stay the same (what if we change the structure later?).
To decouple the components, we should think each part as independently as possible.